langchain/memory/chat_message_histories/momento.py

```diff
@@ -0,0 +1,170 @@
+from __future__ import annotations
+
+import json
+from datetime import timedelta
+from typing import Any, Optional
+
+from langchain.schema import (
+    AIMessage,
+    BaseChatMessageHistory,
+    BaseMessage,
+    HumanMessage,
+    _message_to_dict,
+    messages_from_dict,
+)
+
+
+class MomentoChatMessageHistory(BaseChatMessageHistory):
+    """Chat message history cache that uses Momento as a backend.
+    See https://gomomento.com/"""
+
+    def __init__(
+        self,
+        session_id: str,
+        cache_client: Any,
+        cache_name: str,
+        key_prefix: str = "message_store:",
+        ttl: Optional[timedelta] = None,
+        ensure_cache_exists: bool = True,
+    ):
+        """Instantiate a chat message history cache that uses Momento as a backend.
+
+        Note: to instantiate the cache client passed to MomentoChatMessageHistory,
+        you must have a Momento account at https://gomomento.com/.
+
+        Args:
+            session_id (str): The session ID to use for this chat session.
+            cache_client (CacheClient): The Momento cache client.
+            cache_name (str): The name of the cache to use to store the messages.
+            key_prefix (str, optional): The prefix to apply to the cache key.
+                Defaults to "message_store:".
+            ttl (Optional[timedelta], optional): The TTL to use for the messages.
+                Defaults to None, ie the default TTL of the cache will be used.
+            ensure_cache_exists (bool, optional): Create the cache if it doesn't exist.
+                Defaults to True.
+
+        Raises:
+            ImportError: Momento python package is not installed.
+        """
+        try:
+            from momento import CacheClient
+            from momento.requests import CollectionTtl
+        except ImportError:
+            raise ImportError(
+                "Could not import momento python package. "
+                "Please install it with `pip install momento`."
+            )
+
+        self.cache_client: CacheClient = cache_client
+        self.cache_name = cache_name
+        self.session_id = session_id
+        self.key_prefix = key_prefix
+        if ttl is not None:
+            self.ttl = CollectionTtl.of(ttl)
```

Comment:
just checking, is this intentionally different from how ttl is stored in cache?

Reply:
Good question -- the Momento cache has a separate way of specifying TTL for collections (lists, dictionaries, sets, etc) to distinguish a TTL on creation vs upsert. `CollectionTtl` encapsulates this extra logic.

Reply:
makes sense, thanks!